File: Recursion Maze/Rolling Hash/Q1. Shortest Palindrome/Q1_ShortestPalindrome.py
class Solution(object):
    def shortestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        # if string is empty or has one char, it's already a palindrome
        if not s or len(s) == 1:
            return s

        # reverse the string
        rev = s[::-1]

        # join original and reversed with a separator
        combined = s + "#" + rev

        # get the LPS array for combined string
        lps = self.buildLPS(combined)

        # longest palindrome prefix length
        longest_pal_prefix = lps[-1]

        # find part to add in front from reversed string
        add = rev[:len(s) - longest_pal_prefix]

        # return the final shortest palindrome
        return add + s

    def buildLPS(self, pattern):
        """
        Helper function to build LPS (Longest Prefix Suffix) array
        """
        n = len(pattern)           # length of string
        lps = [0] * n              # create LPS array
        length = 0                 # track prefix length
        i = 1                      # start from second char

        # loop to fill LPS array
        while i < n:
            if pattern[i] == pattern[length]:  # if chars match
                length += 1                    # increase length
                lps[i] = length                # store in LPS
                i += 1                         # move forward
            else:
                if length > 0:                 # if mismatch, try shorter prefix
                    length = lps[length - 1]
                else:                          # if no match
                    lps[i] = 0                 # set 0
                    i += 1                     # move forward
        return lps                             # return final LPS array

File: Recursion Maze/Rolling Hash/Q1. Shortest Palindrome/Q1_ShortestPalindrome.java
class Solution {
    public String shortestPalindrome(String s) {
        if (s == null || s.length() <= 1) return s;   // if empty or single char, it's already palindrome

        String rev = new StringBuilder(s).reverse().toString(); // reverse the string
        String combined = s + "#" + rev;              // combine both with a separator
        int[] lps = buildLPS(combined);               // build KMP prefix table

        int longestPalPrefix = lps[combined.length() - 1]; // length of longest palindromic prefix
        String add = rev.substring(0, rev.length() - longestPalPrefix); // part to add in front
        return add + s;                               // return final shortest palindrome
    }

    // function to build LPS (Longest Prefix Suffix) array like in KMP algorithm
    private int[] buildLPS(String str) {
        int n = str.length();     // length of combined string
        int[] lps = new int[n];   // LPS array
        int len = 0;              // length of previous longest prefix suffix
        int i = 1;                // start from second character

        while (i < n) {           // loop till end
            if (str.charAt(i) == str.charAt(len)) {  // if chars match
                len++;                               // increase length
                lps[i] = len;                        // store in LPS
                i++;                                 // move ahead
            } else {                                 // if chars don't match
                if (len > 0) {                       // try previous prefix
                    len = lps[len - 1];
                } else {                             // no prefix found
                    lps[i] = 0;                      // set 0
                    i++;                             // move ahead
                }
            }
        }
        return lps;               // return the LPS array
    }
}
